File: Harness.EntityFramework5-Beta1/BasicConfiguration.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Threading.Tasks;
using StaticVoid.OrmPerformance.Harness.Contract;

namespace StaticVoid.OrmPerformance.Harness.EntityFramework5_Beta1
{
    public class BasicConfiguration : IRunnableInsertConfiguration, IRunnableUpdateConfiguration, IRunnableSelectConfiguration
    {
        public string Name { get { return "Basic Configuration"; } }

        public string Technology { get { return "Entity Framework 5.0.0.0-Beta1"; } }

        private TestContext _context = null;

        private IConnectionString _connectionString;
        public BasicConfiguration(IConnectionString connectionString)
        {
            _connectionString = connectionString;
        }

        public void Setup()
        {
            _context = new TestContext(_connectionString);
        }

        public void Add(Models.TestEntity entity)
        {
            _context.TestEntities.Add(entity);
        }

        public void Update(int id, string testString, int testInt, DateTime testDateTime)
        {
            var entity = _context.TestEntities.Single(t => t.Id == id);
            entity.TestDate = testDateTime;
            entity.TestInt = testInt;
            entity.TestString = testString;
        }

        public void Commit()
        {
            _context.SaveChanges();
        }
        public void TearDown()
        {
            _context.Dispose();
        }

        public IEnumerable<Models.TestEntity> FindWhereTestIntIs(int testInt)
        {
            return _context.TestEntities.Where(t => t.TestInt == testInt).ToArray();
        }

        public Models.TestEntity Find(int id)
        {
            return _context.TestEntities.SingleOrDefault(t => t.Id == id);
        }
    }
}

File: Harness.EFTypeProvider4/BasicConfiguration.fs
namespace StaticVoid.OrmPerformance.Harness.EFTypeProvider4

open System
open System.Data.Entity
open Microsoft.FSharp.Linq
open StaticVoid.OrmPerformance.Harness.Contract

type DbEntity = TestDb.ServiceTypes.TestEntity
type ModelEntity = StaticVoid.OrmPerformance.Harness.Models.TestEntity

type BasicConfiguration(connectionString:IConnectionString) =

    let mutable _context : TestDb.ServiceTypes.SimpleDataContextTypes.EntityContainer = null

    interface IRunableOrmConfiguration with
        member x.Name       = "Basic Configuration"
        member x.Technology = "Entity Framework Type Provider (EF 4.3.1)"
        
        member x.Setup() =
            _context <- TestDb.GetConfiguredContext(connectionString)

        member x.TearDown() =
            (_context :> IDisposable).Dispose()  
    
    interface ICommitConfiguration with
        member x.Commit() =
            _context.DataContext.SaveChanges() |> ignore

    interface IRunnableInsertConfiguration with

        member x.Add entity =
            DbEntity(
                TestInt = entity.TestInt,
                TestString = entity.TestString,
                TestDate = entity.TestDate
            ) |> _context.TestEntities.AddObject

    interface IRunnableUpdateConfiguration with

        member x.Update(id, testString, testInt, testDateTime) = 
            let entity =
                query {
                    for t in _context.TestEntities do
                    where (t.Id = id)
                    exactlyOne
                }

            entity.TestDate   <- testDateTime
            entity.TestInt    <- testInt
            entity.TestString <- testString

    interface IRunnableDiscreteSelectConfiguration with

        member x.Find id =
            let found =
                query {
                    for t in _context.TestEntities do
                    where (t.Id = id)
                    exactlyOneOrDefault
                }

            if Object.ReferenceEquals(found, null) then
                null
            else
                ModelEntity(
                    TestInt = found.TestInt,
                    TestDate = found.TestDate,
                    TestString = found.TestString
                )

    interface IRunnableBulkSelectConfiguration with

        member x.FindWhereTestIntIs testInt =
            upcast
                (query {
                    for t in _context.TestEntities do
                    where (t.TestInt = testInt)
                    select t }
                |> Seq.map (fun t -> ModelEntity(TestInt = t.TestInt, TestDate = t.TestDate, TestString = t.TestString))
                |> Array.ofSeq)

